### PR TITLE
[mantine/core] popper on Select 

### LINKED ISSUE #202: Add popper.js behavior for Select dropdown
Package: @mantine/core

When select appears on bottom of the page the dropdown is shown outside of viewport.

As discussed on discord this is non-trivial and has accessibility implications.

image

### PR DESCRIPTION
**Pull request by EmilMalanczak**
this PR resolves #202 

Solution seems to work without issues but test are failing I cannot figure out how to fix them.

### CODE REVIEW

File: src/mantine-core/src/components/Autocomplete/Autocomplete.tsx [hunk 1]
```diff
@@ -136,6 +136,9 @@ export const Autocomplete = forwardRef<HTMLInputElement, AutocompleteProps>(
     const { mergedStyles, rest } = useExtractedMargins({ others, style });
     const [dropdownOpened, _setDropdownOpened] = useState(initiallyOpened);
     const [hovered, setHovered] = useState(-1);
+    const [direction, setDirection] = useState<'column' | 'column-reverse'>('column');
```
Comment (EmilMalanczak) [inline src/mantine-core/src/components/Autocomplete/Autocomplete.tsx:139]: I'm not satisfied with that typing, how can I make it shareable - it comes from SelectDropdown component

File: src/mantine-core/src/components/Select/Select.test.tsx [hunk 2]
```diff
@@ -63,9 +63,12 @@ describe('@mantine/core/Select', () => {
     'Select'
   );
 
-  it('renders correct amount of items based on data prop', () => {
-    const element = shallow(<Select data={data.slice(0, 5)} initiallyOpened />);
-    expect(element.render().find('.mantine-Select-item')).toHaveLength(5);
+  it('renders correct amount of items based on data prop', async () => {
+    const element = shallow(<Select {...defaultProps} data={data.slice(0, 5)} initiallyOpened />);
+
+    setTimeout(() => {
```
Comment (EmilMalanczak) [inline src/mantine-core/src/components/Select/Select.test.tsx:69]: that 'hack' makes all of the test passing except `itSupportsStylesApi` and `checkAccessibility`

also I noticed this error:
image

tried to fix it according to this [issue](https://github.com/popperjs/react-popper/issues/350) but no results

File: src/mantine-core/src/components/Select/SelectDropdown/SelectDropdown.tsx [hunk 3]
```diff
@@ -33,37 +37,78 @@ export const SelectDropdown = forwardRef<HTMLDivElement, SelectDropdownProps>(
...
         transition={transition}
-        duration={transitionDuration}
-        timingFunction={transitionTimingFunction}
+        transitionDuration={transitionDuration}
+        transitionTimingFunction={transitionTimingFunction}
+        position="bottom"
+        modifiers={[
+          {
+            name: 'flip',
+            enabled: Boolean(onDirectionChange),
```
Comment (EmilMalanczak) [inline src/mantine-core/src/components/Select/SelectDropdown/SelectDropdown.tsx:60]: should it be passed down to `Popper` component and boolean control prop `withFlip` or something like this?

File: src/mantine-core/src/components/Select/SelectDropdown/SelectDropdown.tsx [hunk 4]
```diff
@@ -33,37 +37,78 @@ export const SelectDropdown = forwardRef<HTMLDivElement, SelectDropdownProps>(
...
+            fn: ({ state }) => {
+              const nextDirection = state.placement === 'top' ? 'column-reverse' : 'column';
+
+              if (direction !== nextDirection) {
+                onDirectionChange && onDirectionChange(nextDirection);
+              }
+            },
+          },
+          {
+            // @ts-ignore
```
Comment (EmilMalanczak) [inline src/mantine-core/src/components/Select/SelectDropdown/SelectDropdown.tsx:79]: ignores due I couldn't find a way to extend built-in type of `react-popper`

**Review by EmilMalanczak — COMMENTED** [review]
Looking forward to feedback

File: src/mantine-core/src/components/Select/Select.test.tsx (hunk 2 above)
Comment (rtivital) [reply src/mantine-core/src/components/Select/Select.test.tsx:69]: Unfortunately we cannot support styles API and some other tests for components that use Popper, you should remove them and hope for the best

File: src/mantine-core/src/components/Select/SelectDropdown/SelectDropdown.tsx (hunk 3 above)
Comment (rtivital) [reply src/mantine-core/src/components/Select/SelectDropdown/SelectDropdown.tsx:60]: I think we should split this into two props:
- onDirectionChange
- allowDirectionChange (defaults to true)

File: src/mantine-core/src/components/Select/SelectDropdown/SelectDropdown.tsx (hunk 4 above)
Comment (rtivital) [reply src/mantine-core/src/components/Select/SelectDropdown/SelectDropdown.tsx:79]: That's fine

File: src/mantine-core/src/components/Autocomplete/Autocomplete.tsx (hunk 1 above)
Comment (rtivital) [reply src/mantine-core/src/components/Autocomplete/Autocomplete.tsx:139]: I think you can use `string` here or `React.CSSProperties['flexDirection']`

**Comment by rtivital** [comment]
Please also resolve conflicts after 3.1.6 patch release

**Comment by rtivital** [comment]
Works great! Do you need to finish something, or I can merge it?

**Comment by EmilMalanczak** [comment]
It's done, feel free to merge :smile:

**Pull request merged**